rpcserver.go

```diff
@@ -2193,13 +2194,104 @@ func (r *rpcServer) AddInvoice(ctx context.Context,
 		options = append(options, zpay32.CLTVExpiry(uint64(defaultDelta)))
 	}
 
+	// If we were requested to include routing hints in the invoice, then
+	// we'll fetch all of our available private channels and create routing
+	// hints for them.
+	if invoice.Private {
+		openChannels, err := r.server.chanDB.FetchAllChannels()
+		if err != nil {
+			return nil, fmt.Errorf("could not fetch all channels")
+		}
+
+		graph := r.server.chanDB.ChannelGraph()
+
+		var routeHints [][]routing.HopHint
+		for _, channel := range openChannels {
+			// We'll restrict the number of individual route hints
+			// to 20 to avoid creating overly large invoices.
+			if len(routeHints) > 20 {
+				break
+			}
+
+			// Since we're only interested in our private channels,
+			// we'll skip public ones.
+			isPublic := channel.ChannelFlags&lnwire.FFAnnounceChannel != 0
+			if isPublic {
+				continue
+			}
+
+			// Make sure the counterparty has enough balance for our
+			// amount.
+			if amtMSat >= channel.LocalCommitment.RemoteBalance {
+				continue
+			}
+
+			// Make sure the channel is active. To do so, we'll need
+			// to satisfy the following:
+			//	1. the counterparty is online
+			//	2. the channel can forward payments
+			var peerOnline, linkActive bool
+			_, err = r.server.FindPeer(channel.IdentityPub)
+			if err == nil {
+				peerOnline = true
+			}
+
+			channelID := lnwire.NewChanIDFromOutPoint(
+				&channel.FundingOutpoint,
+			)
+			link, err := r.server.htlcSwitch.GetLink(channelID)
+			if err == nil {
+				linkActive = link.EligibleToForward()
+			}
+
+			isActive := peerOnline && linkActive
+			if !isActive {
+				continue
+			}
+
+			// Fetch the policies for each end of the channel.
+			chanID := channel.ShortChanID.ToUint64()
+			_, p1, p2, err := graph.FetchChannelEdgesByID(chanID)
+			if err != nil {
+				continue
+			}
+
+			// Now, we'll need to determine which is the correct
+			// policy for HTLCs being sent from the remote node.
+			var remotePolicy *channeldb.ChannelEdgePolicy
+
+			remotePub := channel.IdentityPub.SerializeCompressed()
+			if bytes.Equal(remotePub, p1.Node.PubKeyBytes[:]) {
```

Comment:
This should be `bytes.Compare`, and actually make a decision based on the _ordering_ (not equality) of the two pubkeys.